Packages: Drop shimpei oci add hook

sky1122 opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**

Closes #453 

**Description of changes:**

Replace Shimpei and the OCI-add-hooks packages with NVIDIA Container Runtime to handle OCI hook injection.


**Testing done:**
based on the current branch, running MIG test on k8s-1.32-nvidia variant.
```
{
  "os": {
    "arch": "x86_64",
    "build_id": "e5cd08cf-dirty",
    "pretty_name": "Bottlerocket OS 1.36.0 (aws-k8s-1.32-nvidia)",
    "variant_id": "aws-k8s-1.32-nvidia",
    "version_id": "1.36.0"
  }
}
```

Below is some of the pods
```
╰$ kubectl get pods -A
NAMESPACE     NAME                             READY   STATUS              RESTARTS      AGE
default       nvidia-cuda-workload-45j4d       0/1     ContainerCreating   0             10s
default       nvidia-cuda-workload-4tsv8       0/1     ContainerCreating   0             10s
default       nvidia-cuda-workload-67q25       0/1     ContainerCreating   0             10s
```

```
╰$ kubectl logs nvidia-cuda-workload-45j4d
[Vector addition of 50000 elements]
Copy input data from the host memory to the CUDA device
CUDA kernel launch with 196 blocks of 256 threads
Copy output data from the CUDA device to the host memory
Test PASSED
Done

╰$ kubectl logs nvidia-cuda-workload-4tsv8
[Vector addition of 50000 elements]
Copy input data from the host memory to the CUDA device
CUDA kernel launch with 196 blocks of 256 threads
Copy output data from the CUDA device to the host memory
Test PASSED
Done

 kubectl logs nvidia-cuda-workload-6ggq4
[Vector addition of 50000 elements]
Copy input data from the host memory to the CUDA device
CUDA kernel launch with 196 blocks of 256 threads
Copy output data from the CUDA device to the host memory
Test PASSED
Done
```
Time Slicing and MIG cross setting looks fine
```
Allocatable:
  cpu:                    95690m
  ephemeral-storage:      15751651301
  hugepages-1Gi:          0
  hugepages-2Mi:          0
  memory:                 1167612760Ki
  nvidia.com/gpu:         0
  nvidia.com/gpu.shared:  224
```
Running the CUDA work load and also pass all the test.




- `aws-ecs-2-nvidia`
```
{
    "events": [
        {
            "timestamp": 1743733927639,
            "message": "[Vector addition of 50000 elements]",
            "ingestionTime": 1743733927891
        },
        {
            "timestamp": 1743733927639,
            "message": "Copy input data from the host memory to the CUDA device",
            "ingestionTime": 1743733927891
        },
        {
            "timestamp": 1743733927639,
            "message": "CUDA kernel launch with 196 blocks of 256 threads",
            "ingestionTime": 1743733927891
        },
        {
            "timestamp": 1743733927639,
            "message": "Copy output data from the CUDA device to the host memory",
            "ingestionTime": 1743733927891
        },
        {
            "timestamp": 1743733927639,
            "message": "Test PASSED",
            "ingestionTime": 1743733927891
        },
        {
            "timestamp": 1743733927639,
            "message": "Done",
            "ingestionTime": 1743733927891
        }
    ],
    "nextForwardToken": "f/38886566014310374304316267989737695526165333852819750917/s",
    "nextBackwardToken": "b/38886566014310374304316267989737695526165333852819750912/s"
}
(END)
```
- `aws-k8s-1.32-nvidia`
```
╭─ in ~/k8s
╰$ cat pod-nvidia-test.yaml
apiVersion: v1
kind: Pod
metadata:
  name: k8s-1.32-drop-shimpei
spec:
  restartPolicy: OnFailure
  containers:
    - name: test
      image: nvcr.io/nvidia/k8s/cuda-sample:vectoradd-cuda10.2
      command: ["sleep", "infinity"]
      resources:
        limits:
          nvidia.com/gpu: 1
╭─in ~/k8s
╰$ kubectl apply -f pod-nvidia-test.yaml

╭─ in ~/k8s
╰$ kubectl exec -it k8s-1.32-drop-shimpei -- /bin/bash -c "/tmp/vectorAdd"
[Vector addition of 50000 elements]
Copy input data from the host memory to the CUDA device
CUDA kernel launch with 196 blocks of 256 threads
Copy output data from the CUDA device to the host memory
Test PASSED
Done


Bottlerocket host's root filesystem.
[root@admin]# sheltie
bash-5.1# nvidia-smi
Thu Apr  3 22:44:37 2025
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.230.02             Driver Version: 535.230.02   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  Tesla T4                       On  | 00000000:00:1E.0 Off |                    0 |
| N/A   22C    P8               8W /  70W |      2MiB / 15360MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+

+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|  No running processes found                                                           |
+---------------------------------------------------------------------------------------+
bash-5.1# exit
exit
[root@admin]# apiclient get os
{
  "os": {
    "arch": "x86_64",
    "build_id": "833cf9ad-dirty",
    "pretty_name": "Bottlerocket OS 1.35.0 (aws-k8s-1.32-nvidia)",
    "variant_id": "aws-k8s-1.32-nvidia",
    "version_id": "1.35.0"
  }
}
```

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

arnaldo2792 commented on packages/libnvidia-container/libnvidia-container.spec:
```diff
@@ -17,7 +17,6 @@ Source2: libnvidia-container-sysctl.conf
 Patch0001: 0001-use-shared-libtirpc.patch
 Patch0002: 0002-use-prefix-from-environment.patch
 Patch0003: 0003-keep-debug-symbols.patch
-Patch0004: 0004-Use-NVIDIA_PATH-to-look-up-binaries.patch
```
You should change the number sequence here, and rebuild the patches without this patch, as they were created in the same patch set. By dropping this patch, the patch set must be updated as a whole.

sky1122 replied:
Thank you for pointing out and let me know the best practice on changing patches. Changed.

arnaldo2792 commented on packages/containerd/containerd-config-toml_k8s_containerd_sock:
```diff
@@ -37,15 +37,14 @@ enable_unprivileged_icmp = {{settings.container-runtime.enable-unprivileged-icmp
 
 [plugins."io.containerd.grpc.v1.cri".containerd]
-default_runtime_name = "shimpei"
+default_runtime_name = "runc"
```
nit: I think we can avoid this since the default runtime is `runc`

sky1122 replied:
removed.

arnaldo2792 commented on packages/os/os.spec:
```diff
@@ -21,7 +21,6 @@ Source3: eni-max-pods
 Source6: metricdog-toml
 Source7: host-ctr-toml
-Source8: oci-default-hooks-json
 Source9: cfsignal-toml
```
This sucks, but please renumber the sources.

sky1122 replied:
I noticed the previous source also doesn't get ordered, so I create a separate commit to order the source.

arnaldo2792 replied:
I just saw that you added this as part of this commit, but you should keep this change in the other commit where you changed `os.spec` since that's where the file was actually removed.

arnaldo2792 commented on packages/os/os.spec:
```diff
@@ -698,7 +690,7 @@ if [ -s "%{_cross_repo_root_json}" ] ; then
 
 install -d %{buildroot}%{_cross_templatedir}
-install -p -m 0644 %{S:5} %{S:6} %{S:7} %{S:8} %{S:14} %{S:15} %{S:16} %{S:17} %{S:18} %{S:19} %{S:21} \
+install -p -m 0644 %{S:5} %{S:6} %{S:7} %{S:14} %{S:15} %{S:16} %{S:17} %{S:18} %{S:19} %{S:21} \
```
If `S:8` isn't installed anymore, you have to remove the file from the sources as well. That file is `oci-default-hooks-json`

arnaldo2792 commented on packages/nvidia-container-toolkit/nvidia-container-toolkit.spec:
```diff
@@ -58,6 +58,7 @@ export GOLDFLAGS="-compressdwarf=false -linkmode=external -extldflags '${CGO_LDF
 go build -ldflags="${GOLDFLAGS}" -o nvidia-container-runtime-hook ./cmd/nvidia-container-runtime-hook
 go build -ldflags="${GOLDFLAGS}" -o nvidia-ctk ./cmd/nvidia-ctk
 go build -ldflags="${GOLDFLAGS}" -o nvidia-cdi-hook ./cmd/nvidia-cdi-hook
+go build -ldflags="${GOLDFLAGS}" -o nvidia-container-runtime ./cmd/nvidia-container-runtime
```
I would prefer if you keep adding `nvidia-container-runtime` as its own commit, instead of making it part of the commit that removes `oci-add-hook`

sky1122 commented:
Force Pushed to change the git commit histories to separate drop shimpei and oci-add-hook, change commit message etc.

sky1122 commented:
forced pushed to
- move os.spec changes to drop oci-add-hooks commit
- move install nvidia container runtime to a separate commit

notice something in drop shimpei, will do another forced pushed to take care of that change.

sky1122 commented:
force pushed(s) to
- move some drop oci-add-hooks from shimpei commit (previous one)
- delete default-oci-hooks.json (current one)

arnaldo2792 commented on packages/os/oci-default-hooks-json:
```diff
@@ -1,9 +0,0 @@
-+++
```
This shouldn't be in this commit, you removed this file from a spec in the commit where you drop shimpei

sky1122 committed:
nvidia-container-toolkit: provide nvidia-container-runtime

The nvidia-container-runtime injects the required OCI hooks
that otherwise are provided through shimpei and oci-add-hooks.

Signed-off-by: Jingwei Wang <jweiw@amazon.com>

sky1122 commented:
force pushed to move changes in different commits

notice there is a extra line added, will do another force pushed to remove these from last commit

arnaldo2792 commented on packages/os/os.spec:
```diff
@@ -21,19 +21,18 @@ Source3: eni-max-pods
+Source8: cfsignal-toml
+Source9: warm-pool-wait-toml
+Source10: bottlerocket-cis-checks-metadata-json
+Source11: kubernetes-cis-checks-metadata-json
+Source12: certdog-toml
+Source13: prairiedog-toml
```
Nice, you fixed the numbering (missing 12)! Thanks!

arnaldo2792 commented on packages/os/os.spec:
```diff
@@ -698,7 +689,7 @@ if [ -s "%{_cross_repo_root_json}" ] ; then
 
 install -d %{buildroot}%{_cross_templatedir}
-install -p -m 0644 %{S:5} %{S:6} %{S:7} %{S:8} %{S:14} %{S:15} %{S:16} %{S:17} %{S:18} %{S:19} %{S:21} \
+install -p -m 0644 %{S:5} %{S:6} %{S:7} %{S:12} %{S:13} %{S:14} %{S:15} %{S:16} %{S:17} %{S:19} \
```
You are missing `%{S:18}`.

arnaldo2792 replied:
I was pointed out that line 651 has `S:18`

arnaldo2792 reviewed (changes requested):
You forgot to remove `packages/oci-add-hooks`

sky1122 committed:
libnvidia-container: drop unnecessary patch

The NVIDIA helper binaries are now provided on a standard
location, so this patch isn't needed anymore.

Signed-off-by: Jingwei Wang <jweiw@amazon.com>

sky1122 committed:
packages: drop oci-add-hooks

Signed-off-by: Jingwei Wang <jweiw@amazon.com>

sky1122 commented:
forced pushed to reorder the commit sequent.
forced pushed to pick up the missing commit during the reordering process.

sky1122 commented:
Forced push the commit to reword `nvidia-container-toolkit: provide nvidia-container-runtime` commit

sky1122 commented:
Forced pushed to add sign off in each commits

sky1122 commented:
based on the current branch, running MIG test on k8s-1.32-nvidia variant.
```
{
  "os": {
    "arch": "x86_64",
    "build_id": "e5cd08cf-dirty",
    "pretty_name": "Bottlerocket OS 1.36.0 (aws-k8s-1.32-nvidia)",
    "variant_id": "aws-k8s-1.32-nvidia",
    "version_id": "1.36.0"
  }
}
```

Below is some of the pods
```
╰$ kubectl get pods -A
NAMESPACE     NAME                             READY   STATUS              RESTARTS      AGE
default       nvidia-cuda-workload-45j4d       0/1     ContainerCreating   0             10s
default       nvidia-cuda-workload-4tsv8       0/1     ContainerCreating   0             10s
default       nvidia-cuda-workload-67q25       0/1     ContainerCreating   0             10s
```

```
╰$ kubectl logs nvidia-cuda-workload-45j4d
[Vector addition of 50000 elements]
Copy input data from the host memory to the CUDA device
CUDA kernel launch with 196 blocks of 256 threads
Copy output data from the CUDA device to the host memory
Test PASSED
Done

╰$ kubectl logs nvidia-cuda-workload-4tsv8
[Vector addition of 50000 elements]
Copy input data from the host memory to the CUDA device
CUDA kernel launch with 196 blocks of 256 threads
Copy output data from the CUDA device to the host memory
Test PASSED
Done

 kubectl logs nvidia-cuda-workload-6ggq4
[Vector addition of 50000 elements]
Copy input data from the host memory to the CUDA device
CUDA kernel launch with 196 blocks of 256 threads
Copy output data from the CUDA device to the host memory
Test PASSED
Done
```
Time Slicing and MIG cross setting looks fine
```
Allocatable:
  cpu:                    95690m
  ephemeral-storage:      15751651301
  hugepages-1Gi:          0
  hugepages-2Mi:          0
  memory:                 1167612760Ki
  nvidia.com/gpu:         0
  nvidia.com/gpu.shared:  224
```
Running the CUDA work load and also pass all the test.

arnaldo2792 commented on packages/nvidia-container-toolkit/nvidia-container-toolkit.spec:
```diff
@@ -16,15 +16,13 @@ URL: https://%{goimport}
 Source0: https://%{goimport}/archive/v%{gover}/nvidia-container-toolkit-%{gover}.tar.gz
 Source1: nvidia-container-toolkit-config-k8s.toml
 Source2: nvidia-container-toolkit-config-ecs.toml
-Source3: nvidia-oci-hooks-json
```
I asked @sky1122 to keep this change and the removal of `nvidia-oci-hooks-json` in this commit and not in the other commit that modifies `nvidia-container-toolkit` because removing this file was used by `shimpei`.

arnaldo2792 commented:
I cloned these changes and confirmed that bypassing GPU requests through `NVIDIA_VISIBLE_DEVICES=all` in k8s variants work as expected. The same test in ECS doesn't have any effect because the ECS agent clobbers the `NVIDIA_VISIBLE_DEVICES` with the actual list of the devices allocated for the ECS tasks.

bcressey commented on packages/containerd/containerd-config-toml_k8s_containerd_sock:
```diff
@@ -36,16 +36,12 @@ enable_unprivileged_ports = {{settings.container-runtime.enable-unprivileged-por
-default_runtime_name = "shimpei"
-
-[plugins."io.containerd.grpc.v1.cri".containerd.runtimes.shimpei]
+[plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runc]
```
Do we still need to specify `default_runtime_name` somewhere, or does it default to `runc`?

arnaldo2792 replied:
`containerd` defaults to `runc`

bcressey commented on packages/os/os.spec:
```diff
@@ -21,19 +21,18 @@ Source3: eni-max-pods
-Source19: host-containers-toml
-Source20: bottlerocket-fips-checks-metadata-json
-Source21: bootstrap-commands-toml
+Source8: cfsignal-toml
```
Personally, I'd opt to leave a gap in source files vs. renumbering them all, in the interests of making it easier to review.

It's the same rationale behind not using spaces to vertically align columns, and ending lists with a trailing `,` in contexts where that's supported. Don't add more "green" to the review if you can help it, unless it's in service of a technical goal beyond mere cosmetics.

It may seem innocuous but what you've done here is caused the reviewer to have to verify that every line in `%install` that copies one of these files has been modified correctly, which requires a lot of scrolling back and forth.

sky1122 replied:
Thanks for sharing the idea behind these. 
Change the os.spec to only contain drop shimpei and oci-add-hooks in the forced push.

bcressey reviewed (commented):
Please minimize the changes to `nvidia-container-toolkit.spec` and `os.spec` and then I'll take another look. Most of it looks fine to me but it's not necessary to do a lot of source renumbering.

sky1122 committed:
packages: drop shimpei

Signed-off-by: Jingwei Wang <jweiw@amazon.com>
